File: ios/Classes/Wrapper/Commands/Config/ChangeApplicationCodeCommand.swift
//
//  Created by Emarsys on 2021. 08. 10..
//
import EmarsysSDK

class ChangeApplicationCodeCommand: EmarsysCommandProtocol {
    private var pushEventHandler: EMSEventHandlerBlock
    private var silentPushEventHandler: EMSEventHandlerBlock
    private var geofenceEventHandler: EMSEventHandlerBlock
    private var inAppEventHandler: EMSEventHandlerBlock

    init(pushEventHandler: @escaping EMSEventHandlerBlock,
         silentPushEventHandler: @escaping EMSEventHandlerBlock,
         geofenceEventHandler: @escaping EMSEventHandlerBlock,
         inAppEventHandler: @escaping EMSEventHandlerBlock) {
        self.pushEventHandler = pushEventHandler
        self.silentPushEventHandler = silentPushEventHandler
        self.geofenceEventHandler = geofenceEventHandler
        self.inAppEventHandler = inAppEventHandler
    }

    func execute(arguments: [String : Any]?, resultCallback: @escaping ResultCallback) {
        guard let applicationCode = arguments?["applicationCode"] as? String else {
            resultCallback(["error": "Invalid applicationCode"])
            return
        }
        Emarsys.config.changeApplicationCode(applicationCode: applicationCode) { error in
            if let e = error {
                resultCallback(["error": e])
            } else {
                Emarsys.push.silentMessageEventHandler = self.silentPushEventHandler
                Emarsys.push.notificationEventHandler = self.pushEventHandler
                Emarsys.geofence.eventHandler = self.geofenceEventHandler
                Emarsys.inApp.eventHandler = self.inAppEventHandler
                
                resultCallback(["success": true])
            }
        }
    }
}

File: android/src/main/kotlin/com/emarsys/emarsys_sdk/command/config/ChangeApplicationCodeCommand.kt
package com.emarsys.emarsys_sdk.command.config

import android.content.SharedPreferences
import com.emarsys.Emarsys
import com.emarsys.emarsys_sdk.command.EmarsysCommand
import com.emarsys.emarsys_sdk.command.ResultCallback
import com.emarsys.emarsys_sdk.config.ConfigStorageKeys
import com.emarsys.emarsys_sdk.di.dependencyContainer
import androidx.core.content.edit

class ChangeApplicationCodeCommand(private val sharedPreferences: SharedPreferences) :
    EmarsysCommand {
    override fun execute(parameters: Map<String, Any?>?, resultCallback: ResultCallback) {
        val applicationCode = parameters?.get("applicationCode") as String?
        if (applicationCode != null) {
            Emarsys.config.changeApplicationCode(applicationCode) {
                if (it == null) {
                    sharedPreferences.edit {
                        putString(
                            ConfigStorageKeys.MOBILE_ENGAGE_APPLICATION_CODE.name, applicationCode
                        )
                    }
                    registerEventHandlers()
                } else {
                    sharedPreferences.edit {
                        remove(ConfigStorageKeys.MOBILE_ENGAGE_APPLICATION_CODE.name)
                    }
                }
                resultCallback.invoke(null, it)
            }
        } else {
            resultCallback(null, IllegalArgumentException("applicationCode must not be null"))
        }
    }

    private fun registerEventHandlers() {
        Emarsys.push.setNotificationEventHandler(dependencyContainer().pushEventHandler)
        Emarsys.push.setSilentMessageEventHandler(dependencyContainer().silentPushEventHandler)
        Emarsys.inApp.setEventHandler(dependencyContainer().inAppEventHandler)
        Emarsys.geofence.setEventHandler(dependencyContainer().geofenceEventHandler)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        return true
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }
}